File: backend/app/schemas/mata_kuliah.py
"""
backend/app/schemas/mata_kuliah.py
Pydantic v2 schemas untuk entitas MataKuliah dan MataKuliahKelas.
"""

import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict


class MataKuliahCreate(BaseModel):
    kode: str
    kurikulum_id: uuid.UUID
    nama: str
    sks: int
    semester: int
    jenis: str  # Wajib / Pilihan
    prasyarat: Optional[str] = None


class MataKuliahUpdate(BaseModel):
    kode: Optional[str] = None
    kurikulum_id: Optional[uuid.UUID] = None
    nama: Optional[str] = None
    sks: Optional[int] = None
    semester: Optional[int] = None
    jenis: Optional[str] = None
    prasyarat: Optional[str] = None


class KurikulumNested(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    kode: str
    tahun: str
    prodi_id: uuid.UUID


class MataKuliahResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    kode: str
    kurikulum_id: uuid.UUID
    nama: str
    sks: int
    semester: int
    jenis: str
    prasyarat: Optional[str]
    is_active: bool
    created_at: datetime
    kurikulum: Optional[KurikulumNested] = None


# ---------------------------------------------------------------------------
# MataKuliahKelas schemas
# ---------------------------------------------------------------------------

class MataKuliahKelasCreate(BaseModel):
    kelas: Optional[str] = None
    label: str
    ket: Optional[str] = None


class MataKuliahKelasUpdate(BaseModel):
    kelas: Optional[str] = None
    label: Optional[str] = None
    ket: Optional[str] = None


class MataKuliahKelasResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    mata_kuliah_id: uuid.UUID
    kelas: Optional[str]
    label: str
    ket: Optional[str]
    created_at: datetime

File: frontend/src/api/mataKuliah.ts
import apiClient from './client'

// ─── Types ────────────────────────────────────────────────────────────────────

export interface KurikulumNested {
  id: string
  kode: string
  tahun: string
  prodi_id: string
}

export interface MataKuliah {
  id: string
  kode: string
  kurikulum_id: string
  nama: string
  sks: number
  semester: number
  jenis: string
  prasyarat: string | null
  is_active: boolean
  created_at: string
  kurikulum: KurikulumNested | null
}

export interface MataKuliahCreatePayload {
  kode: string
  kurikulum_id: string
  nama: string
  sks: number
  semester: number
  jenis: string
  prasyarat: string | null
}

export interface MataKuliahKelas {
  id: string
  mata_kuliah_id: string
  kelas: string | null
  label: string
  ket: string | null
  created_at: string
}

export interface MataKuliahKelasPayload {
  kelas: string | null
  label: string
  ket: string | null
}

export interface Kurikulum {
  id: string
  kode: string
  tahun: string
  prodi_id: string
  is_active: boolean
}

// ─── Mata Kuliah API ──────────────────────────────────────────────────────────

export interface MataKuliahFilter {
  prodi_id?: string
  kurikulum_id?: string
  semester?: number
}

export async function getMataKuliah(filters?: MataKuliahFilter): Promise<MataKuliah[]> {
  const params: Record<string, string | number> = {}
  if (filters?.prodi_id) params.prodi_id = filters.prodi_id
  if (filters?.kurikulum_id) params.kurikulum_id = filters.kurikulum_id
  if (filters?.semester !== undefined) params.semester = filters.semester
  const res = await apiClient.get<MataKuliah[]>('/mata-kuliah', { params })
  return res.data
}

export async function createMataKuliah(data: MataKuliahCreatePayload): Promise<MataKuliah> {
  const res = await apiClient.post<MataKuliah>('/mata-kuliah', data)
  return res.data
}

export async function updateMataKuliah(
  id: string,
  data: Partial<MataKuliahCreatePayload>
): Promise<MataKuliah> {
  const res = await apiClient.put<MataKuliah>(`/mata-kuliah/${id}`, data)
  return res.data
}

export async function deleteMataKuliah(id: string): Promise<void> {
  await apiClient.delete(`/mata-kuliah/${id}`)
}

// ─── Kelas API ────────────────────────────────────────────────────────────────

export async function getKelas(mkId: string): Promise<MataKuliahKelas[]> {
  const res = await apiClient.get<MataKuliahKelas[]>(`/mata-kuliah/${mkId}/kelas`)
  return res.data
}

export async function createKelas(
  mkId: string,
  data: MataKuliahKelasPayload
): Promise<MataKuliahKelas> {
  const res = await apiClient.post<MataKuliahKelas>(`/mata-kuliah/${mkId}/kelas`, data)
  return res.data
}

export async function updateKelas(
  mkId: string,
  kelasId: string,
  data: Partial<MataKuliahKelasPayload>
): Promise<MataKuliahKelas> {
  const res = await apiClient.put<MataKuliahKelas>(
    `/mata-kuliah/${mkId}/kelas/${kelasId}`,
    data
  )
  return res.data
}

export async function deleteKelas(mkId: string, kelasId: string): Promise<void> {
  await apiClient.delete(`/mata-kuliah/${mkId}/kelas/${kelasId}`)
}

// ─── Kurikulum API ────────────────────────────────────────────────────────────

export async function getKurikulum(): Promise<Kurikulum[]> {
  const res = await apiClient.get<Kurikulum[]>('/kurikulum')
  return res.data
}
